server/routes/choropleth.py

```diff
@@ -0,0 +1,215 @@
+                    (not per_capita or
+                    'data' in population_by_geo.get(geo_id, []))):
+                # Grab the latest available data.
+                stat_obs = next(iter(
+                    reversed(stat_var_by_geo[geo_id]['data'].values())))
+
+                # Handle per-capita option.
+                if per_capita:
+                    # Note that this could be from a different period from the
+                    # statistical variable.
+                    # TODO(iancostello): Determine how to handle populations 
+                    # and statistical values from different times.
+                    pop = next(iter(
+                        reversed(population_by_geo[geo_id]['data'].values())))
+                    stat_obs /= pop
+
+                values.append(stat_obs)
+                geo_feature["properties"]["value"] = stat_obs
+
+            # Add to main dataframe.
+            features.append(geo_feature)
+    # Generate plotting information from value list.
+    domain, palette = determine_color_palette(values, per_capita)
+
+    # Return as json payload.
+    return jsonify({
+        "geoJson": {
+            "type": "FeatureCollection",
+            "features": features
+        },
+        "_PLOTTING_INFO": {
+            "domain": domain,
+            "palette": palette
+        }  
+    }, 200)
+
+def coerce_geojson_to_righthand_rule(geoJsonCords, obj_type):
+    """Changes GeoJSON handedness to the right-hand rule.
+    
+    GeoJSON is stored in DataCommons in the reverse format of what D3
+    expects. This results in geographies geometry being inverted. This function
+    fixes these lists to be in the format expected by D3 and turns all polygons
+    into multipolygons for downstream consistency.
+        Args:
+            geoJsonCords -> Nested list of geojson.
+            obj_type -> Object feature type.
+        Returns:
+            Nested list of geocoords.
+    """
+    if obj_type == "Polygon":
+        geoJsonCords[0].reverse()
+        return [geoJsonCords]
+    elif obj_type == "MultiPolygon":
+        for polygon in geoJsonCords:
+            polygon[0].reverse()
+        return geoJsonCords
+    else:
+        assert False, f"Type {obj_type} unknown!"
+
+def determine_color_palette(values, is_denominated):
+    """Builds plotting range and color domain given a set of values.
+        Percentage values use diverging palettes.
+        Sequential data uses sequential palettes.
+
+    Args:
+        values -> List of all float values in the choropleth.
+        is_denominated -> Whether the requested data has a measurement 
+            denominator like per capita, as a boolean. 
+    Returns:
+        scale_type, lower_domain, upper_domain, palette.
+    """
+    # Percentages use diverging palettes.
+    if is_denominated:
+        lower_range = np.quantile(values, 0.10)
+        upper_range = np.quantile(values, 0.90)
+        palette = ['#998ec3', '#f7f7f7', '#f1a340']
```

Comment:
where are these colors taken from? d3 has fantastic color libraries.
can this computation move to the client? i'm also unsure how the quantile computation here translate into different palettes.

Reply:
From a site dedicated to Choropleth coloring. Will add a todo.